karapace/kafka_rest_apis/__init__.py

```diff
@@ -134,37 +146,217 @@ def _add_kafka_rest_routes(self) -> None:
             method="GET",
             rest_request=True,
+            with_request=True,
+            json_body=False,
         )
         self.route(
             "/topics/<topic:path>/partitions/<partition_id:path>",
             callback=self.partition_publish,
             method="POST",
             rest_request=True,
+            with_request=True,
+        )
+        self.route(
+            "/topics/<topic:path>/partitions",
+            callback=self.list_partitions,
+            method="GET",
+            rest_request=True,
+            with_request=True,
+            json_body=False,
         )
-        self.route("/topics/<topic:path>/partitions", callback=self.list_partitions, method="GET", rest_request=True)
         # Topics
-        self.route("/topics", callback=self.list_topics, method="GET", rest_request=True)
-        self.route("/topics/<topic:path>", callback=self.topic_details, method="GET", rest_request=True)
-        self.route("/topics/<topic:path>", callback=self.topic_publish, method="POST", rest_request=True)
+        self.route("/topics", callback=self.list_topics, method="GET", rest_request=True, with_request=True, json_body=False)
+        self.route(
+            "/topics/<topic:path>",
+            callback=self.topic_details,
+            method="GET",
+            rest_request=True,
+            with_request=True,
+            json_body=False,
+        )
+        self.route("/topics/<topic:path>", callback=self.topic_publish, method="POST", rest_request=True, with_request=True)
+
+    def get_user_proxy(self, request: HTTPRequest) -> "UserRestProxy":
+        log.info("get user proxy for %s", request)
+        key = ""
+        if self.config.get("rest_authorization", False):
+            log.info("Using rest authorization")
+            auth_header = request.headers.get("Authorization")
+            log.info("Using rest auth %s", auth_header)
+
+            if auth_header is None:
+                raise HTTPResponse(
+                    body='{"message": "Unauthorized"}',
+                    status=HTTPStatus.UNAUTHORIZED,
+                    content_type=JSON_CONTENT_TYPE,
+                    headers={"WWW-Authenticate": 'Basic realm="Karapace REST Proxy"'},
+                )
+            key = auth_header
+            if self.proxies.get(key) is None:
+                auth = aiohttp.BasicAuth.decode(auth_header)
+                config = self.config.copy()
+                config["bootstrap_uri"] = config["sasl_bootstrap_uri"]
+                config["security_protocol"] = (
+                    "SASL_SSL" if config["security_protocol"] in ("SSL", "SASL_SSL") else "SASL_PLAINTEXT"
+                )
+                config["sasl_mechanism"] = "PLAIN"
+                config["sasl_plain_username"] = auth.login
+                config["sasl_plain_password"] = auth.password
+                self.proxies[key] = UserRestProxy(config, self.kafka_timeout, self.serializer)
```

Comment:
What about the following.

Change `self.proxies` like this:

```python
self.proxies: Dict[str, Dict[str, "UserRestProxy"]] = {}
```

The outer key is the username and the inner key is the password.

I think we don't ever need to store the "old" proxy when the user changes his password, or do we? 

If not, `get_user_proxy` should work like this.

* If there's a proxy for the username & password tuple, return that
* If there's no proxy for the username, create and store a new one and return that
* If there's a proxy for the username, but for a different password, "close" the proxy and create a new like in the above case.

Old proxies will use the old password and should not be able to connect to Kafka, though not sure how Kafka authentication works for existing connections when the password is changed. But no matter how it works, we _shouldn't_ talk to Kafka with old credentials.

Reply:
Maybe `Dict[Username, UserRestProxy]`:
```suggestion
maybe_valid_proxy = proxies.get(username)
if maybe_valid_proxy is None:
    # If proxy is None a new must be created
    proxy = create_user_proxy(...)
elif maybe_valid_proxy.get_pw_hash() != hash(auth.password):
    # If password has changed, close the old, create new
    maybe_valid_proxy.close()
    proxy = create_user_proxy(...)
    proxies[username] = proxy
else:
    # existing will do
    proxy = maybe_valid_proxy
```

Reply:
At least cannot close old proxy unless authentication with new password succeeds, otherwise malicious calls cause havoc.

But even for normal usage, closing old proxies automatically also causes destruction of consumers & subscriptions in uncontrolled fashion. At least I haven't found ultimate solution for password change cases, which would not cause some oddities one way or another.